elements/dash-tab-panel/dash-tab-panel.js

```diff
@@ -11,167 +11,216 @@
-    // Calling `appendChild` on an already inserted element _moves_ the
-    // element to the last child position.
-    tabs.forEach(tab => this.appendChild(tab));
-    panels.forEach(panel => this.appendChild(panel));
-
-    // The element checks if any of the tabs have been marked as selected. If
-    // not, the first tab is now selected.
-    const selectedTab =
-      tabs.find(e => e.getAttribute('aria-selected') === 'true') || tabs[0];
-
-    // Next, we switch to the selected tab. `selectTab` takes care of marking
-    // all other tabs as deselected and hiding all other panels.
-    this._selectTab(selectedTab);
-  }
+  class TabList extends HTMLElement {
+
+    /**
+     * The constructor hooks up the event handlers. Since only events on the
+     * element itself are handled, this can take place in the constructor.
+     */
+    constructor() {
+      super();
+    }
 
-  /**
-   * `reset` marks all tabs as deselected and hides all the panels.
-   */
-  reset() {
-    const tabs = Array.from(this.querySelectorAll('[aria-role=tab]'));
-    const panels = Array.from(this.querySelectorAll('[aria-role=tabpanel]'));
-
-    tabs.forEach(tab => {
-      tab.classList.remove('selected');
-      tab.setAttribute('tabindex', '-1');
-      tab.setAttribute('aria-selected', 'false');
-    });
-
-    panels.forEach(panel => {
-      panel.classList.add('hidden');
-      panel.setAttribute('aria-hidden', 'true');
-    });
-  }
+    /**
+     * `connectedCallback` groups tabs and panels by reordering and makes sure
+     * exactly one tab is active.
+     */
+    connectedCallback() {
+      // The element needs to do some manual input event handling to allow
+      // switching with arrow keys and Home/End.
+      this.addEventListener('keydown', this._onKeyDown);
+      this.addEventListener('click', this._onClick);
+
+      this.setAttribute('role', 'tablist');
+      // Acquire all tabs and panels inside the element
+      const tabs = Array.from(this.querySelectorAll('[role=tab]'));
+      const panels = Array.from(this.querySelectorAll('[role=tabpanel]'));
+      // If there are no tabs, there is no way to switch between panels. Abort.
+      if (!tabs.length) return;
+
+      // For progressive enhancement, the markup should alternate between tabs
+      // and panels. If JavaScript is disabled, all panels are
+      // visible with their respective tab right above them.
+      // If JavaScript is enabled, the element reorders all children into
+      // their two groups. First all the tabs, then all the panels.
+      // Calling `appendChild` on an already inserted element _moves_ the
+      // element to the last child position.
+      tabs.forEach(tab => this.appendChild(tab));
+      panels.forEach(panel => this.appendChild(panel));
+
+      // Give each panel a `aria-labelledby` attribute that refers to the tab
+      // that controls it.
+      tabs.forEach(tab => {
+        const panelId = tab.getAttribute('aria-controls');
```

Comment:
Yeah I get that. My hope is we can show folks that they can use ARIA in a "modern" fashion where the element is smart and self applies it. That way folks don't feel like they have to spray their hand written markup with attributes which can be hard to remember.

A possible idea is we do a second pass in the future and really explain what the ARIA attributes are doing in the JS. So rather than just saying something like "foo() sets aria-checked to true" we could spend a bit of time in the comment diving into what it means for aria-checked to be true for assistive technology and the accessibility tree. I'll have to think on that a bit more :)